packages/liveblocks-core/src/lib/fsm.ts

```diff
@@ -28,51 +28,48 @@ export type AsyncErrorEvent = {
 export type BaseEvent = { readonly type: string };
 export type BuiltinEvent = TimerEvent | AsyncOKEvent<unknown> | AsyncErrorEvent;
 
-export type CleanupFn = () => void;
+export type Patchable<TContext> = Readonly<TContext> & {
+  patch(patch: Partial<TContext>): void;
+};
+
+export type CleanupFn<TContext> = (context: Patchable<TContext>) => void;
 export type EnterFn<TContext> = (
-  context: Readonly<TContext>
-) => void | CleanupFn;
+  context: Patchable<TContext>
+) => void | CleanupFn<TContext>;
 
 export type TargetFn<
-  TContext,
+  TContext extends object,
   TEvent extends BaseEvent,
   TState extends string
 > = (
   event: TEvent,
   context: Readonly<TContext>
 ) => TState | TargetConfig<TContext, TEvent, TState> | null;
 
-export type Assigner<TContext, TEvent extends BaseEvent> =
-  | Partial<TContext>
-  | ((context: Readonly<TContext>, event: TEvent) => Partial<TContext>);
-
 export type Effect<TContext, TEvent extends BaseEvent> = (
-  context: Readonly<TContext>,
+  context: Patchable<TContext>,
   event: TEvent
 ) => void;
 
 export type TargetConfig<
-  TContext,
+  TContext extends object,
   TEvent extends BaseEvent,
   TState extends string
 > = {
   target: TState;
 
-  /**
-   * Specify an object that will be used to "patch" the current context as soon
-   * as the transition is taken. The context will be updated before the new
-   * state is entered.
-   */
-  assign?: Assigner<TContext, TEvent>;
-
   /**
    * Emit a side effect (other than assigning to the context) when this
    * transition is taken.
    */
-  effect?: Effect<TContext, TEvent>;
+  effect: Effect<TContext, TEvent> | Effect<TContext, TEvent>[];
```

Comment:
does this mean we require an effect now? Maybe an empty array at least?

Reply:
Yes, but only if you use the "expanded form", i.e. `{ target: "@my-state", effect: <something> }`. If you wouldn't want to use an effect, you'd typically use the shorter string form, i.e. `"@my-state"`.